README.md
<a name="readme-top"></a>

[JA](README.md) | [EN](README.en.md)

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![License][license-shield]][license-url]

# Azure Kinect ROS Driver

<!-- 目次 -->
<details>
  <summary>目次</summary>
  <ol>
    <li>
      <a href="#概要">概要</a>
    </li>
    <li>
      <a href="#環境構築">環境構築</a>
      <ul>
        <li><a href="#環境条件">環境条件</a></li>
        <li><a href="#インストール方法">インストール方法</a></li>
      </ul>
    </li>
    <li>
    　<a href="#実行操作方法">実行・操作方法</a>
      <ul>
        <li><a href="#動作確認">動作確認</a></li>
        <li><a href="#Rviz上の可視化">Rviz上の可視化</a></li>
        <li><a href="#デモ骨格検出">デモ：骨格検出</a></li>
        <li><a href="#デモRtabMap">デモ：RtabMap</a></li>
      </ul>
    </li>
    <li>
    <li><a href="#マイルストーン">マイルストーン</a></li>
    <!-- <li><a href="#contributing">Contributing</a></li> -->
    <!-- <li><a href="#license">License</a></li> -->
    <li><a href="#参考文献">参考文献</a></li>
  </ol>
</details>



<!-- レポジトリの概要 -->
## 概要

Microsoftが作成した[Azure Kinect ROS Driver](https://github.com/microsoft/Azure_Kinect_ROS_Driver)を基に，`ROS Noetic`にも対応させたレポジトリです．
さらに，実際のカメラの特性に忠実で，更新された`URDF`ファイルをサポートしています．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>



<!-- セットアップ -->
## セットアップ

ここで，本レポジトリのセットアップ方法について説明します．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


### 環境条件

まず，以下の環境を整えてから，次のインストール段階に進んでください．

| System  | Version |
| ------------- | ------------- |
| Ubuntu | 20.04 (Focal Fossa) |
| ROS | Noetic Ninjemys |

> [!NOTE]
> `Ubuntu`や`ROS`のインストール方法に関しては，[SOBIT Manual](https://github.com/TeamSOBITS/sobits_manual#%E9%96%8B%E7%99%BA%E7%92%B0%E5%A2%83%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6)に参照してください．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


### インストール方法

1. ROSの`src`フォルダに移動します．
   ```sh
   $ roscd
   # もしくは，"cd ~/catkin_ws/"へ移動．
   $ cd src/
   ```
2. 本レポジトリをcloneします．
   ```sh
   $ git clone https://github.com/TeamSOBITS/azure_kinect_ros_driver
   ```
3. レポジトリの中へ移動します．
   ```sh
   $ cd azure_kinect_ros_driver/
   ```
4. 依存パッケージをインストールします．
   ```sh
   $ bash install.sh
   ```

> [!NOTE]
> [install.sh](install.sh)を実行することによって，必要とされる[Azure Kinect SDK (K4A)](https://github.com/microsoft/Azure-Kinect-Sensor-SDK)が自動的にインストールされます．

5. パッケージをコンパイルします．
   ```sh
   $ roscd
   # もしくは，"cd ~/catkin_ws/"へ移動．
   $ catkin_make
   ```

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


<!-- 実行・操作方法 -->
## 実行・操作方法

カメラ単体で使用する場合は，[driver.launch](launch/driver.launch)を次の手順で実行してください．


1. 必要に応じて[driver.launch](launch/driver.launch)のパラメータを更新してください．
   ```xml
    <!-- Add Prefix to tf frame. It typically contains a trailing '_' unless empty. -->
    <arg name="tf_prefix"         default="azure_" />
    <!-- Add cover to camera -->
    <arg name="add_cover"         default="true" />
    <!-- Visualize Rviz -->
    <arg name="open_rviz"         default="true"/>
    ...
   ```

> [!NOTE]
> 使用したい機能に応じて，`true`か`false`かに書き換えてください．

2. [driver.launch](launch/driver.launch)というlaunchファイルを実行します．
   ```sh
   $ roslaunch azure_kinect_ros_driver driver.launch
   ```

> [!NOTE]
> カメラのパラメータを変更する場合は，[usage.md](docs/usage.md)を参照してください．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


### 動作確認

Azure Kinect ROS Driverが正常にインストールされたかどうかを確認するために，[install.sh](install.sh)でインストールされた`k4a-tools`の`k4aviewer`を実行してみます．

1. `k4aviewer`を起動します．
    ```sh
    $ k4aviewer
    ```

2. `Open Device`という項目の中に，カメラを指定します．

> [!NOTE]
> `Azure Kinect`が認識されない場合があります．その時は，`Refresh Devices`というボタンを押してみてください．問題が直せない場合は，[ライトの意味](https://learn.microsoft.com/ja-jp/azure/kinect-dk/hardware-specification#what-does-the-light-mean)を参照にして，USBの接続を確認してみてください．

3. カメラと接続ができた場合，必要に応じてカメラのパラメータを更新してください．

4. 最後に，`Connect`ボタンを押し，カメラを起動します．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


### Rviz上の可視化
実機を動かす前段階として，Rviz上でAzure Kinect DKを可視化し，カメラの構成を表示することができます．

```sh
$ roslaunch azure_kinect_ros_driver display.launch
```

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


### デモ：骨格検出

Azure Kinect ROS Driverがデフォルトとして用意している骨格検出のデモプログラムを実行できます．

1. [driver_with_bodytracking.launch](launch/driver_with_bodytracking.launch)というlaunchファイルを実行します．
    ```sh
    $ roslaunch azure_kinect_ros_driver driver_with_bodytracking.launch
    ```

> [!NOTE]
> [実行・操作方法](#実行・操作方法)の項目に示されているように，[driver.launch](launch/driver.launch)というlaunchファイルを必要に応じてパラメータを修正してください．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


### デモ：RtabMap

Azure Kinect ROS Driverがデフォルトとして用意しているRtabMapのデモプログラムを実行できます．

1. [slam_rtabmap.launch](launch/slam_rtabmap.launch)というlaunchファイルを実行します．
    ```sh
    $ roslaunch azure_kinect_ros_driver slam_rtabmap.launch
    ```

> [!NOTE]
> [実行・操作方法](#実行・操作方法)の項目に示されているように，必要に応じてパラメータを修正してください．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


<!-- マイルストーン -->
## マイルストーン

- [x] OSS
    - [x] ドキュメンテーションの充実
    - [x] [install.sh](install.sh)の更新

現時点のバッグや新規機能の依頼を確認するために[Issueページ][license-url] をご覧ください．

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>


<!-- CONTRIBUTING -->
<!-- ## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#readme-top">上に戻る</a>)</p> -->


<!-- LICENSE -->
<!-- ## License

Distributed under the MIT License. See `LICENSE.txt` for more NOTErmation.

<p align="right">(<a href="#readme-top">上に戻る</a>)</p> -->


<!-- 参考文献 -->
## 参考文献

* [Azure Kinect DK](https://azure.microsoft.com/ja-jp/products/kinect-dk)
* [Azure Kinect SDK (K4A)](https://github.com/microsoft/Azure-Kinect-Sensor-SDK)
* [Azure Kinect ROS Driver](https://github.com/microsoft/Azure_Kinect_ROS_Driver)
* [Azure Kinect DK のドキュメント](https://learn.microsoft.com/ja-jp/azure/kinect-dk/)
* [ROS Noetic](http://wiki.ros.org/noetic)

<p align="right">(<a href="#readme-top">上に戻る</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/TeamSOBITS/azure_kinect_ros_driver.svg?style=for-the-badge
[contributors-url]: https://github.com/TeamSOBITS/azure_kinect_ros_driver/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/TeamSOBITS/azure_kinect_ros_driver.svg?style=for-the-badge
[forks-url]: https://github.com/TeamSOBITS/azure_kinect_ros_driver/network/members
[stars-shield]: https://img.shields.io/github/stars/TeamSOBITS/azure_kinect_ros_driver.svg?style=for-the-badge
[stars-url]: https://github.com/TeamSOBITS/azure_kinect_ros_driver/stargazers
[issues-shield]: https://img.shields.io/github/issues/TeamSOBITS/azure_kinect_ros_driver.svg?style=for-the-badge
[issues-url]: https://github.com/TeamSOBITS/azure_kinect_ros_driver/issues
[license-shield]: https://img.shields.io/github/license/TeamSOBITS/azure_kinect_ros_driver.svg?style=for-the-badge
[license-url]: LICENSE


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package sobits_common
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (202X-XX-XX)
-------------------
* first public release for ROS Noetic

install.sh
#!/bin/bash
# Reference: https://github.com/microsoft/Azure-Kinect-Sensor-SDK/issues/1263#issuecomment-710698591

echo "╔══╣ Install: k4a-tools for Azure Kinect (STARTING) ╠══╗"


# Install dependencies
sudo apt-get update
sudo apt-get install -y \
    software-properties-common \
    curl \
    wget \
    dialog \
    libsoundio1

# Download Azure Kinect related packages
curl -O https://packages.microsoft.com/ubuntu/18.04/prod/pool/main/libk/libk4a1.4/libk4a1.4_1.4.1_amd64.deb
curl -O https://packages.microsoft.com/ubuntu/18.04/prod/pool/main/libk/libk4a1.4-dev/libk4a1.4-dev_1.4.1_amd64.deb
curl -O https://packages.microsoft.com/ubuntu/18.04/prod/pool/main/libk/libk4abt1.1/libk4abt1.1_1.1.2_amd64.deb
curl -O https://packages.microsoft.com/ubuntu/18.04/prod/pool/main/libk/libk4abt1.1-dev/libk4abt1.1-dev_1.1.2_amd64.deb
curl -O https://packages.microsoft.com/ubuntu/18.04/prod/pool/main/k/k4a-tools/k4a-tools_1.4.1_amd64.deb

# Install k4a-related packages
echo 'libk4a1.4 libk4a1.4/accepted-eula-hash string 0f5d5c5de396e4fee4c0753a21fee0c1ed726cf0316204edda484f08cb266d76' | sudo debconf-set-selections
echo 'libk4a1.4 libk4a1.4/accept-eula boolean true' | sudo debconf-set-selections
sudo dpkg -i libk4a1.4_1.4.1_amd64.deb
sudo dpkg -i libk4a1.4-dev_1.4.1_amd64.deb

echo 'libk4abt1.1 libk4abt1.1/accepted-eula-hash string 03a13b63730639eeb6626d24fd45cf25131ee8e8e0df3f1b63f552269b176e38' | sudo debconf-set-selections
echo 'libk4abt1.1 libk4abt1.1/accept-eula boolean true' | sudo debconf-set-selections
sudo dpkg -i libk4abt1.1_1.1.2_amd64.deb
sudo dpkg -i libk4abt1.1-dev_1.1.2_amd64.deb

sudo dpkg -i k4a-tools_1.4.1_amd64.deb

# Remove deb files
sudo rm *.deb

# Replace k4aConfig.cmake
sudo sed -i 's/find_dependency(k4a 1.4 REQUIRED)/find_dependency(k4a 1.4.1 REQUIRED)/g' /usr/lib/cmake/k4abt/k4abtConfig.cmake

# Add rules file
wget https://raw.githubusercontent.com/microsoft/Azure-Kinect-Sensor-SDK/develop/scripts/99-k4a.rules
sudo mv 99-k4a.rules /etc/udev/rules.d/

# Install ROS packages for rtabmap demo
sudo apt-get install -y \
    ros-${ROS_DISTRO}-rtabmap-ros \
    ros-${ROS_DISTRO}-camera-info-manager


echo "╚══╣ Install: k4a-tools for Azure Kinect (FINISHED) ╠══╝"


src/k4a_calibration_transform_data.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// Associated header
//
#include "azure_kinect_ros_driver/k4a_calibration_transform_data.h"

// System headers
//
#include <stdexcept>

// Library headers
//
#include <angles/angles.h>
#include <sensor_msgs/distortion_models.h>
#include <tf2/LinearMath/Transform.h>
#include <tf2/convert.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>

// Project headers
//
#include "azure_kinect_ros_driver/k4a_ros_types.h"

void K4ACalibrationTransformData::initialize(const k4a::device& device, const k4a_depth_mode_t depth_mode,
                                             const k4a_color_resolution_t resolution, const K4AROSDeviceParams params)
{
  k4a_calibration_ = device.get_calibration(depth_mode, resolution);
  initialize(params);
}

void K4ACalibrationTransformData::initialize(const k4a::playback& k4a_playback_handle, const K4AROSDeviceParams params)
{
  k4a_calibration_ = k4a_playback_handle.get_calibration();
  initialize(params);
}

void K4ACalibrationTransformData::initialize(const K4AROSDeviceParams params)
{
  k4a_transformation_ = k4a::transformation(k4a_calibration_);
  tf_prefix_          = params.tf_prefix;
  // camera_base_frame_  = params.camera_base_frame;
  // rgb_camera_frame_   = params.rgb_camera_frame;
  // depth_camera_frame_ = params.depth_camera_frame;
  // imu_frame_          = params.imu_frame;

  print();

  bool depthEnabled = (getDepthWidth() * getDepthHeight() > 0);
  bool colorEnabled = (getColorWidth() * getColorHeight() > 0);

  // Create a buffer to store the point cloud
  if (params.point_cloud && (!params.rgb_point_cloud || params.point_cloud_in_depth_frame))
  {
    point_cloud_image_ = k4a::image::create(K4A_IMAGE_FORMAT_DEPTH16, getDepthWidth(), getDepthHeight(),
                                            getDepthWidth() * 3 * (int) sizeof(DepthPixel));
  }
  else if (params.point_cloud && params.rgb_point_cloud)
  {
    point_cloud_image_ = k4a::image::create(K4A_IMAGE_FORMAT_DEPTH16, getColorWidth(), getColorHeight(),
                                            getColorWidth() * 3 * (int) sizeof(DepthPixel));
  }

  if (depthEnabled && colorEnabled)
  {
    // Create a buffer to store RGB images that are transformed into the depth camera geometry
    transformed_rgb_image_ = k4a::image::create(K4A_IMAGE_FORMAT_COLOR_BGRA32, getDepthWidth(), getDepthHeight(),
                                                getDepthWidth() * (int) sizeof(BgraPixel));

    // Create a buffer to store depth images that are transformed into the RGB camera geometry
    transformed_depth_image_ = k4a::image::create(K4A_IMAGE_FORMAT_DEPTH16, getColorWidth(), getColorHeight(),
                                                  getColorWidth() * (int) sizeof(DepthPixel));
  }

  // Publish various transforms needed by ROS.
  // We publish all TFs all the time, even if the respective sensor data
  // output is off. This allows us to just use the SDK calibrations, and one
  // cosmetic TF output between the depth and the base.
  publishDepthToBaseTf();
  publishImuToDepthTf();
  publishRgbToDepthTf();
}

int K4ACalibrationTransformData::getDepthWidth() { return k4a_calibration_.depth_camera_calibration.resolution_width; }

int K4ACalibrationTransformData::getDepthHeight()
{
  return k4a_calibration_.depth_camera_calibration.resolution_height;
}

int K4ACalibrationTransformData::getColorWidth() { return k4a_calibration_.color_camera_calibration.resolution_width; }

int K4ACalibrationTransformData::getColorHeight()
{
  return k4a_calibration_.color_camera_calibration.resolution_height;
}

void K4ACalibrationTransformData::print()
{
  ROS_INFO("K4A Calibration Blob:");
  ROS_INFO("\t Depth:");
  printCameraCalibration(k4a_calibration_.depth_camera_calibration);

  ROS_INFO("\t Color:");
  printCameraCalibration(k4a_calibration_.color_camera_calibration);

  ROS_INFO("\t IMU (Depth to Color):");
  printExtrinsics(k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_DEPTH][K4A_CALIBRATION_TYPE_COLOR]);

  ROS_INFO("\t IMU (Depth to IMU):");
  printExtrinsics(k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_DEPTH][K4A_CALIBRATION_TYPE_ACCEL]);

  ROS_INFO("\t IMU (IMU to Depth):");
  printExtrinsics(k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_ACCEL][K4A_CALIBRATION_TYPE_DEPTH]);

  ROS_INFO("\t IMU (Color to IMU):");
  printExtrinsics(k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_COLOR][K4A_CALIBRATION_TYPE_ACCEL]);

  ROS_INFO("\t IMU (IMU to Color):");
  printExtrinsics(k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_ACCEL][K4A_CALIBRATION_TYPE_COLOR]);
}

void K4ACalibrationTransformData::printCameraCalibration(k4a_calibration_camera_t& calibration)
{
  printExtrinsics(calibration.extrinsics);

  ROS_INFO("\t\t Resolution:");
  ROS_INFO_STREAM("\t\t\t Width: " << calibration.resolution_width);
  ROS_INFO_STREAM("\t\t\t Height: " << calibration.resolution_height);

  ROS_INFO("\t\t Intrinsics:");
  ROS_INFO_STREAM("\t\t\t Model Type: " << calibration.intrinsics.type);
  ROS_INFO_STREAM("\t\t\t Parameter Count: " << calibration.intrinsics.parameter_count);
  ROS_INFO_STREAM("\t\t\t cx: " << calibration.intrinsics.parameters.param.cx);
  ROS_INFO_STREAM("\t\t\t cy: " << calibration.intrinsics.parameters.param.cy);
  ROS_INFO_STREAM("\t\t\t fx: " << calibration.intrinsics.parameters.param.fx);
  ROS_INFO_STREAM("\t\t\t fy: " << calibration.intrinsics.parameters.param.fy);
  ROS_INFO_STREAM("\t\t\t k1: " << calibration.intrinsics.parameters.param.k1);
  ROS_INFO_STREAM("\t\t\t k2: " << calibration.intrinsics.parameters.param.k2);
  ROS_INFO_STREAM("\t\t\t k3: " << calibration.intrinsics.parameters.param.k3);
  ROS_INFO_STREAM("\t\t\t k4: " << calibration.intrinsics.parameters.param.k4);
  ROS_INFO_STREAM("\t\t\t k5: " << calibration.intrinsics.parameters.param.k5);
  ROS_INFO_STREAM("\t\t\t k6: " << calibration.intrinsics.parameters.param.k6);
  ROS_INFO_STREAM("\t\t\t codx: " << calibration.intrinsics.parameters.param.codx);
  ROS_INFO_STREAM("\t\t\t cody: " << calibration.intrinsics.parameters.param.cody);
  ROS_INFO_STREAM("\t\t\t p2: " << calibration.intrinsics.parameters.param.p2);
  ROS_INFO_STREAM("\t\t\t p1: " << calibration.intrinsics.parameters.param.p1);
  ROS_INFO_STREAM("\t\t\t metric_radius: " << calibration.intrinsics.parameters.param.metric_radius);
}

void K4ACalibrationTransformData::printExtrinsics(k4a_calibration_extrinsics_t& extrinsics)
{
  ROS_INFO("\t\t Extrinsics:");
  ROS_INFO_STREAM("\t\t\t Translation: " << extrinsics.translation[0] << ", " << extrinsics.translation[1] << ", "
                                         << extrinsics.translation[2]);
  ROS_INFO_STREAM("\t\t\t Rotation[0]: " << extrinsics.rotation[0] << ", " << extrinsics.rotation[1] << ", "
                                         << extrinsics.rotation[2]);
  ROS_INFO_STREAM("\t\t\t Rotation[1]: " << extrinsics.rotation[3] << ", " << extrinsics.rotation[4] << ", "
                                         << extrinsics.rotation[5]);
  ROS_INFO_STREAM("\t\t\t Rotation[2]: " << extrinsics.rotation[6] << ", " << extrinsics.rotation[7] << ", "
                                         << extrinsics.rotation[8]);
}

void K4ACalibrationTransformData::publishRgbToDepthTf()
{
  k4a_calibration_extrinsics_t* rgb_extrinsics =
      &k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_DEPTH][K4A_CALIBRATION_TYPE_COLOR];
  tf2::Vector3 depth_to_rgb_translation(rgb_extrinsics->translation[0] / 1000.0f,
                                        rgb_extrinsics->translation[1] / 1000.0f,
                                        rgb_extrinsics->translation[2] / 1000.0f);
  tf2::Matrix3x3 depth_to_rgb_rotation(
      rgb_extrinsics->rotation[0], rgb_extrinsics->rotation[1], rgb_extrinsics->rotation[2],
      rgb_extrinsics->rotation[3], rgb_extrinsics->rotation[4], rgb_extrinsics->rotation[5],
      rgb_extrinsics->rotation[6], rgb_extrinsics->rotation[7], rgb_extrinsics->rotation[8]);
  tf2::Transform depth_to_rgb_transform(depth_to_rgb_rotation, depth_to_rgb_translation);

  geometry_msgs::TransformStamped static_transform;
  static_transform.transform = tf2::toMsg(depth_to_rgb_transform.inverse());

  static_transform.header.stamp = ros::Time::now();
  static_transform.header.frame_id = tf_prefix_ + depth_camera_frame_;
  static_transform.child_frame_id = tf_prefix_ + rgb_camera_frame_;

  static_broadcaster_.sendTransform(static_transform);
}

void K4ACalibrationTransformData::publishImuToDepthTf()
{
  k4a_calibration_extrinsics_t* imu_extrinsics =
      &k4a_calibration_.extrinsics[K4A_CALIBRATION_TYPE_DEPTH][K4A_CALIBRATION_TYPE_ACCEL];
  tf2::Vector3 depth_to_imu_translation(imu_extrinsics->translation[0] / 1000.0f,
                                        imu_extrinsics->translation[1] / 1000.0f,
                                        imu_extrinsics->translation[2] / 1000.0f);
  tf2::Matrix3x3 depth_to_imu_rotation(
      imu_extrinsics->rotation[0], imu_extrinsics->rotation[1], imu_extrinsics->rotation[2],
      imu_extrinsics->rotation[3], imu_extrinsics->rotation[4], imu_extrinsics->rotation[5],
      imu_extrinsics->rotation[6], imu_extrinsics->rotation[7], imu_extrinsics->rotation[8]);
  tf2::Transform depth_to_imu_transform(depth_to_imu_rotation, depth_to_imu_translation);

  geometry_msgs::TransformStamped static_transform;
  static_transform.transform = tf2::toMsg(depth_to_imu_transform.inverse());

  static_transform.header.stamp = ros::Time::now();
  static_transform.header.frame_id = tf_prefix_ + depth_camera_frame_;
  static_transform.child_frame_id = tf_prefix_ + imu_frame_;

  static_broadcaster_.sendTransform(static_transform);
}

void K4ACalibrationTransformData::publishDepthToBaseTf()
{
  // This is a purely cosmetic transform to make the base model of the URDF look good.
  geometry_msgs::TransformStamped static_transform;

  static_transform.header.stamp = ros::Time::now();
  static_transform.header.frame_id = tf_prefix_ + camera_base_frame_;
  static_transform.child_frame_id = tf_prefix_ + depth_camera_frame_;

  tf2::Vector3 depth_translation = getDepthToBaseTranslationCorrection();
  static_transform.transform.translation.x = depth_translation.x();
  static_transform.transform.translation.y = depth_translation.y();
  static_transform.transform.translation.z = depth_translation.z();

  tf2::Quaternion depth_rotation = getDepthToBaseRotationCorrection();
  static_transform.transform.rotation.x = depth_rotation.x();
  static_transform.transform.rotation.y = depth_rotation.y();
  static_transform.transform.rotation.z = depth_rotation.z();
  static_transform.transform.rotation.w = depth_rotation.w();

  static_broadcaster_.sendTransform(static_transform);
}

// The [0,0,0] center of the URDF, the TF frame known as "camera_base", is offset slightly from the
// [0,0,0] origin of the depth camera frame, known as "depth_camera_link" or "depth_camera_frame"
//
// Publish a TF link so the URDF model and the depth camera line up correctly
#define DEPTH_CAMERA_OFFSET_MM_X 0.0f
#define DEPTH_CAMERA_OFFSET_MM_Y 0.0f
#define DEPTH_CAMERA_OFFSET_MM_Z 1.8f  // The depth camera is shifted 1.8mm up in the depth window

tf2::Vector3 K4ACalibrationTransformData::getDepthToBaseTranslationCorrection()
{
  // These are purely cosmetic tranformations for the URDF drawing!!
  return tf2::Vector3(DEPTH_CAMERA_OFFSET_MM_X / 1000.0f, DEPTH_CAMERA_OFFSET_MM_Y / 1000.0f,
                      DEPTH_CAMERA_OFFSET_MM_Z / 1000.0f);
}

tf2::Quaternion K4ACalibrationTransformData::getDepthToBaseRotationCorrection()
{
  // These are purely cosmetic tranformations for the URDF drawing!!
  tf2::Quaternion ros_camera_rotation;  // ROS camera co-ordinate system requires rotating the entire camera relative to
                                        // camera_base
  tf2::Quaternion depth_rotation;       // K4A has one physical camera that is about 6 degrees downward facing.

  depth_rotation.setEuler(0, angles::from_degrees(-6.0), 0);
  ros_camera_rotation.setEuler(M_PI / -2.0f, M_PI, (M_PI / 2.0f));

  return ros_camera_rotation * depth_rotation;
}

void K4ACalibrationTransformData::getDepthCameraInfo(sensor_msgs::CameraInfo& camera_info)
{
  camera_info.header.frame_id = tf_prefix_ + depth_camera_frame_;
  camera_info.width = getDepthWidth();
  camera_info.height = getDepthHeight();
  camera_info.distortion_model = sensor_msgs::distortion_models::RATIONAL_POLYNOMIAL;

  k4a_calibration_intrinsic_parameters_t* parameters = &k4a_calibration_.depth_camera_calibration.intrinsics.parameters;

  // The distortion parameters, size depending on the distortion model.
  // For "rational_polynomial", the 8 parameters are: (k1, k2, p1, p2, k3, k4, k5, k6).
  camera_info.D = {parameters->param.k1, parameters->param.k2, parameters->param.p1, parameters->param.p2,
                   parameters->param.k3, parameters->param.k4, parameters->param.k5, parameters->param.k6};

  // clang-format off
  // Intrinsic camera matrix for the raw (distorted) images.
  //     [fx  0 cx]
  // K = [ 0 fy cy]
  //     [ 0  0  1]
  // Projects 3D points in the camera coordinate frame to 2D pixel
  // coordinates using the focal lengths (fx, fy) and principal point
  // (cx, cy).
  camera_info.K = {parameters->param.fx,  0.0f,                   parameters->param.cx,
                   0.0f,                  parameters->param.fy,   parameters->param.cy,
                   0.0f,                  0.0,                    1.0f};

  // Projection/camera matrix
  //     [fx'  0  cx' Tx]
  // P = [ 0  fy' cy' Ty]
  //     [ 0   0   1   0]
  // By convention, this matrix specifies the intrinsic (camera) matrix
  //  of the processed (rectified) image. That is, the left 3x3 portion
  //  is the normal camera intrinsic matrix for the rectified image.
  // It projects 3D points in the camera coordinate frame to 2D pixel
  //  coordinates using the focal lengths (fx', fy') and principal point
  //  (cx', cy') - these may differ from the values in K.
  // For monocular cameras, Tx = Ty = 0. Normally, monocular cameras will
  //  also have R = the identity and P[1:3,1:3] = K.
  camera_info.P = {parameters->param.fx,  0.0f,                   parameters->param.cx,   0.0f,
                   0.0f,                  parameters->param.fy,   parameters->param.cy,   0.0f,
                   0.0f,                  0.0,                    1.0f,                   0.0f};

  // Rectification matrix (stereo cameras only)
  // A rotation matrix aligning the camera coordinate system to the ideal
  // stereo image plane so that epipolar lines in both stereo images are
  // parallel.
  camera_info.R = {1.0f, 0.0f, 0.0f,
                   0.0f, 1.0f, 0.0f,
                   0.0f, 0.0f, 1.0f};
  // clang-format on
}

void K4ACalibrationTransformData::getRgbCameraInfo(sensor_msgs::CameraInfo& camera_info)
{
  camera_info.header.frame_id = tf_prefix_ + rgb_camera_frame_;
  camera_info.width = getColorWidth();
  camera_info.height = getColorHeight();
  camera_info.distortion_model = sensor_msgs::distortion_models::RATIONAL_POLYNOMIAL;

  k4a_calibration_intrinsic_parameters_t* parameters = &k4a_calibration_.color_camera_calibration.intrinsics.parameters;

  // The distortion parameters, size depending on the distortion model.
  // For "rational_polynomial", the 8 parameters are: (k1, k2, p1, p2, k3, k4, k5, k6).
  camera_info.D = {parameters->param.k1, parameters->param.k2, parameters->param.p1, parameters->param.p2,
                   parameters->param.k3, parameters->param.k4, parameters->param.k5, parameters->param.k6};

  // clang-format off
  // Intrinsic camera matrix for the raw (distorted) images.
  //     [fx  0 cx]
  // K = [ 0 fy cy]
  //     [ 0  0  1]
  // Projects 3D points in the camera coordinate frame to 2D pixel
  // coordinates using the focal lengths (fx, fy) and principal point
  // (cx, cy).
  camera_info.K = {parameters->param.fx,  0.0f,                   parameters->param.cx,
                   0.0f,                  parameters->param.fy,   parameters->param.cy,
                   0.0f,                  0.0,                    1.0f};

  // Projection/camera matrix
  //     [fx'  0  cx' Tx]
  // P = [ 0  fy' cy' Ty]
  //     [ 0   0   1   0]
  // By convention, this matrix specifies the intrinsic (camera) matrix
  //  of the processed (rectified) image. That is, the left 3x3 portion
  //  is the normal camera intrinsic matrix for the rectified image.
  // It projects 3D points in the camera coordinate frame to 2D pixel
  //  coordinates using the focal lengths (fx', fy') and principal point
  //  (cx', cy') - these may differ from the values in K.
  // For monocular cameras, Tx = Ty = 0. Normally, monocular cameras will
  //  also have R = the identity and P[1:3,1:3] = K.
  camera_info.P = {parameters->param.fx,  0.0f,                   parameters->param.cx,   0.0f,
                   0.0f,                  parameters->param.fy,   parameters->param.cy,   0.0f,
                   0.0f,                  0.0,                    1.0f,                   0.0f};

  // Rectification matrix (stereo cameras only)
  // A rotation matrix aligning the camera coordinate system to the ideal
  // stereo image plane so that epipolar lines in both stereo images are
  // parallel.
  camera_info.R = {1.0f, 0.0f, 0.0f,
                   0.0f, 1.0f, 0.0f,
                   0.0f, 0.0f, 1.0f};
  // clang-format on
}


install_k4a.sh
#!/bin/bash
# Reference: https://github.com/microsoft/Azure-Kinect-Sensor-SDK/issues/1263#issuecomment-710698591

echo "╔══╣ Install: k4a-tools for Azure Kinect (STARTING) ╠══╗"


# Install dependencies
arch=amd64

sudo apt-get update
sudo apt install -y \
    software-properties-common \
    curl \
    dialog

# Add Microsoft’s official GPG key (Use 18.04 for 20.04)
curl -sSL https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
sudo apt-add-repository https://packages.microsoft.com/ubuntu/18.04/prod

# Set up the stable repository
curl -sSL https://packages.microsoft.com/config/ubuntu/18.04/prod.list | sudo tee /etc/apt/sources.list.d/microsoft-prod.list

# Install the latest stable version of k4a
sudo apt-get update

sudo apt-get install -y \
    libk4a1.3-dev \
    libk4abt1.0-dev \
    k4a-tools=1.3.0

# Add rules file
wget https://raw.githubusercontent.com/microsoft/Azure-Kinect-Sensor-SDK/develop/scripts/99-k4a.rules
sudo mv 99-k4a.rules /etc/udev/rules.d/

# Install ROS packages for rtabmap demo
sudo apt-get install -y \
    ros-${ROS_DISTRO}-rtabmap-ros

echo "╚══╣ Install: k4a-tools for Azure Kinect (FINISHED) ╠══╝"
